File: 백준/Gold/1062. 가르침/가르침.cc
#include <bits/stdc++.h>
using namespace std;

int n, m, ret;
string s;
int a[54];

void check(int learn){
    int cnt = 0;
    for(int i=0; i<n; i++){
        if((a[i] & learn) == a[i]) cnt++;
    }
    ret = max(ret,cnt);
    return;
}

void go(int idx, int count, int learn){
    if(count < 0) return;
    if(idx == 26) {
        check(learn);
        return;
    }
    go(idx+1, count-1, learn | (1<<idx));
    if(idx != 'a'-'a' && idx != 'n'-'a' && idx != 't'-'a' && idx != 'i'-'a' && idx != 'c'-'a'){
        go(idx+1,count,learn);
    }
    return;
}
int main(){
    cin >> n >> m;
    for(int i=0; i<n; i++){
        cin >> s;
        for(char str : s){
            a[i] |= (1<<(str-'a'));
        }
    }

    go(0,m,0);

    cout << ret << "\n";
    return 0;
}

// a. 단어마다 하나의 숫자로 저장하고
// b. 단어 학습
//  1. 바로학습
//  2. a,n,t,i,c 제외하 경우는 안배울수있음.
// c. idx == 26인 경우는 z까지 오면서 배울단어를 모두 고른상태에서 check해서 배운것과 단어들의 값을 비교

File: 백준/Gold/1062. 가르침/가르침.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main{

    static int[] words;
    static int n, m, ans;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        ans = -1;

        words = new int[n];
        for(int i=0; i<n; i++){
            int sum = 0;
            char[] chars = br.readLine().toCharArray();
            for(int j=0; j<chars.length; j++){
                sum |= (1 << (chars[j] - 'a'));
            }
            words[i] = sum;
        }

        if(m < 5){
            System.out.println(0);
            return;
        }
        go(0, 0, 0);

        System.out.println(ans);
    }

    static void go(int learn, int count, int word){
        if(count == m || word == 26){
            int ret = 0;
            for(int i=0; i<n; i++){
                if((words[i] & learn) == words[i]){ // 배운 단어라면
                    ret++;
                }
            }
            ans = Math.max(ans, ret);
            return;
        }

        if(word == 0 || word == 2 || word == 8 || word == 13 || word == 19){
            go(learn | (1<<word), count+1, word+1); // a n t i c 면 배운다.
        }
        else{ // 아니라면
            go(learn | (1<<word), count+1, word+1); // 배우거나
            go(learn, count, word+1); // 배우지 않는다.
        }
    }
}